pallets/referrals/src/lib.rs

```diff
@@ -0,0 +1,573 @@
+	#[pallet::storage]
+	#[pallet::getter(fn asset_tier)]
+	pub(super) type AssetTier<T: Config> =
+		StorageDoubleMap<_, Blake2_128Concat, T::AssetId, Blake2_128Concat, Level, Tier, OptionQuery>;
+
+	/// Information about assets that are currently in the rewards pot.
+	/// Used to easily determine list of assets that need to be converted.
+	#[pallet::storage]
+	#[pallet::getter(fn assets)]
+	pub(super) type Assets<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetId, ()>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Referral code has been registered.
+		CodeRegistered {
+			code: ReferralCode<T::CodeLength>,
+			account: T::AccountId,
+		},
+		/// Referral code has been linked to an account.
+		CodeLinked {
+			account: T::AccountId,
+			code: ReferralCode<T::CodeLength>,
+			referral_account: T::AccountId,
+		},
+		/// Asset has been converted to RewardAsset.
+		Converted {
+			from: T::AssetId,
+			to: T::AssetId,
+			amount: Balance,
+			received: Balance,
+		},
+		/// Rewards claimed.
+		Claimed { who: T::AccountId, rewards: Balance },
+		/// New asset tier has been set.
+		TierRewardSet {
+			asset_id: T::AssetId,
+			level: Level,
+			referrer: Permill,
+			trader: Permill,
+		},
+	}
+
+	#[pallet::error]
+	#[cfg_attr(test, derive(PartialEq, Eq))]
+	pub enum Error<T> {
+		/// Referral code is too long.
+		TooLong,
+		/// Referral code is too short.
+		TooShort,
+		/// Referral code contains invalid character. Only alphanumeric characters are allowed.
+		InvalidCharacter,
+		/// Referral code already exists.
+		AlreadyExists,
+		/// Provided referral code is invalid. Either does not exist or is too long.
+		InvalidCode,
+		/// Account is already linked to another referral account.
+		AlreadyLinked,
+		/// Nothing in the referral pot account for the asset.
+		ZeroAmount,
+		/// Linking an account to the same referral account is not allowed.
+		LinkNotAllowed,
+		/// Calculated rewards are more than the fee amount. This can happen if percentages are incorrectly set.
+		IncorrectRewardCalculation,
+		/// Given referrer and trader percentages exceeds 100% percent.
+		IncorrectRewardPercentage,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Register new referral code.
+		///
+		/// `origin` pays the registration fee.
+		/// `code` is assigned to the given `account`.
+		///
+		/// Length of the `code` must be at least `MIN_CODE_LENGTH`.
+		/// Maximum length is limited to `T::CodeLength`.
+		/// `code` must contain only alfa-numeric characters and all characters will be converted to upper case.
+		///
+		/// Parameters:
+		/// - `code`: Code to register. Must follow the restrictions.
+		///
+		/// Emits `CodeRegistered` event when successful.
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::register_code())]
+		pub fn register_code(origin: OriginFor<T>, code: Vec<u8>) -> DispatchResult {
```

Comment:
cannot you limit length of the string on the input as well ?

Reply:
I could used bounded vec for param too. But in the polkadot.js when i tested it - it throws some weird errors when i try to enter longer code. something about wasm unreachable.

Reply:
thats good tho, you shouldn't input longer code :)

Reply:
changed to bounded vec